### PR TITLE
Random cdc

### PR DESCRIPTION
**Pull request by lorenzoCapelli**
Done the implementation of:

1. Random Projection 
2. Class-Dependent Detector

### CODE REVIEW

File: peepholelib/coreVectors/dimReduction/randomProjection.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+# torch stuff
+import torch
+
+# Our stuff
+from peepholelib.models.model_wrap import get_in_activations
+from .dim_reduction_base import DimReductionBase as DRB 
+from torch.utils.data import DataLoader
+
+class RandomProjection(DRB):
+    def __init__(self, **kwargs):
```
Comment (lsrosa) [inline peepholelib/coreVectors/dimReduction/randomProjection.py:30]: documentation missing

File: peepholelib/coreVectors/dimReduction/randomProjection.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        self._random_proj = None
+        self.reduct_m = None       
+
+        if self.file_path.exists():
+            if self.verbose: print(f'File {self.file_path} exists. Loading from disk.')
+            self._random_proj = torch.load(self.file_path)
+            self.reduct_m = self._random_proj['reduct_m'].detach().to(self.device)
+        else:
+
+            first_ds = next(iter(datasets._dss.values()))
```
Comment (lsrosa) [inline peepholelib/coreVectors/dimReduction/randomProjection.py]: you probably should be using the parsed ds here, not the `dataset wrap`.

Also, it would be better to pass a sample from the experiment, so this class does not have to deal with the inner workings of the other classes, like `linear_svd`

File: peepholelib/coreVectors/dimReduction/randomProjection.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            model.set_activations(save_input=save_input, save_output=save_output)
+
+            with torch.no_grad():
+                _in = sample_in.to(self.device)
+                model(_in)
+                _act0 = activations_parser(model._acts)
+            
+            proj_dim = _act0[layer].flatten(start_dim=1).shape[1]
+
+            model.set_activations(save_input=False, save_output=False)
```
Comment (lsrosa) [inline peepholelib/coreVectors/dimReduction/randomProjection.py:68]: `model` is not being used, and the `dim_reduction` should not worry about it and the activations, this is done in `get_corevectors`

File: peepholelib/coreVectors/dimReduction/randomProjection.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            tcvs (torch.tensor): Trimmed corevectors and correspective labels
+            labels (torch.tensor): Labels from datasate for the samples. Only returned if `dss` is given
+        """
+
+        cvs = kwargs['cvs']
+        dss = kwargs.get('dss', None)
+        label_key = kwargs.get('label_key', 'label') 
+
+        # trim corevectors on the last dimension
+        tcvs = cvs[...,0:self.cv_dim]
```
Comment (lsrosa) [inline peepholelib/coreVectors/dimReduction/randomProjection.py:124]: I am not familiar with the `QR` decomposition, does it have the same property as the `SVDs` that `QR(n+1)[0:n] = QR(n)`? otherwise this trimming would not make sense.

File: peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py [hunk 5]
```diff
@@ -24,8 +24,7 @@ def __call__(self, **kwargs):
         # Assuming the input is of shape (batch_size, num_tokens, num_channels)
 
         act_data = kwargs['act_data'] 
-        act_data = act_data[:, 1:, :]  # Exclude the class token
-        cvs = torch.mean(act_data, dim=2)  # Mean across tokens
+        cvs = act_data.mean(dim=(1, 2))
```
Comment (lsrosa) [inline peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py:27]: This seems odd and not equivalent to what we have previously.

I tried this on a python terminal:
```py
x = torch.rand(3, 4, 5)
torch.mean(x[:, 1:, :], dim=2)
x.mean(dim=(1, 2))
```

Is this a fix? Are the dimensions I tested wrong?

File: peepholelib/peepholes/ClassDependentClassifiers/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+__all__ = ['tgmm', 'tkmeans', 'anomaly_detector']
```
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/__init__.py]: this should be unecessary

File: peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+# torch stuff
+import torch
+
+# persistence
+import joblib
+
+
+class AnomalyDetector(CDCBase):
+    '''
+    Class-Dependent Classifier using any sklearn-compatible anomaly detector.
```
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py]: Can you remove these braklines in the documentation lines, it messes up depending on the editor and window size you use, it is better to leave it a single lines and leave the editors to wrap the text.

File: peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        from sklearn.ensemble import IsolationForest
+        cdc = AnomalyDetector(detector_cls=IsolationForest, cls_kwargs={'n_estimators': 100}, ...)
+    '''
+
+    def __init__(self, **kwargs):
+        detector_cls = kwargs['detector_cls']
+        cls_kwargs = kwargs['cls_kwargs']
+        CDCBase.__init__(self, **kwargs)
+
+        if not hasattr(detector_cls(), 'score_samples'):
```
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py]: check also for `fit()`

File: peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+            raise RuntimeError(
+                f'Data has shape {data.shape} after parsing, '
+                f'but n_features={self.n_features} was expected.'
+            )
+
+        detector_name = type(self._classifiers[0]).__name__
+        if verbose: print(f'\n ---- Class-Dependent-Classifier {detector_name} (one per class)\n')
+
+        for c in range(self.nl_model):
+            class_data = data[labels == c].cpu().numpy()
```
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py]: note that this prevents any type of classifier to be used on the GPU. While this works for sklearn, this should not be a general thing. Should probably set a device argument. To solve the `numpy` dependency you need to either add a wrap (complicated at this stage) or add another argument and an `if`

File: peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        Compute score_samples of each sample under each class detector.
+        Higher score = more inlier-like for that class.
+
+        Args:
+        - data (torch.Tensor): Parsed corevectors of shape (N, n_features).
+
+        Returns:
+        - scores (torch.Tensor): Shape (N, nl_model).
+        '''
+        data = kwargs['data'].cpu().numpy()
```
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py]: again, numpy

File: peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py (hunk 7 above)
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py]: Note that sklearn-style modules are not always sklearn, the standard is wildly adopted.

File: peepholelib/peepholes/ClassDependentClassifiers/tgmm.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+import torch
+
+# https://github.com/CSOgroup/torchgmm/tree/main
+from torchgmm.bayes import GaussianMixture as tGMM
+
+import logging
+logging.getLogger('pytorch_lightning.utilities.rank_zero').setLevel(logging.CRITICAL)
+logging.getLogger("pytorch_lightning.accelerators.cuda").setLevel(logging.CRITICAL)
+
+class GMM(CDCBase):
```
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/tgmm.py]: Can we give a better name for this class? `ClassDependentGMM`?

File: peepholelib/peepholes/ClassDependentClassifiers/tgmm.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+            dim=1
+        ) 
+        
+        return log_densities
+
+    def predict(self, data):
+        probs = self.classifier_probabilities(data=data)
+        return probs.argmax(dim=1)
+    
+    def save_single_class(self, **kwargs):
```
Comment (lsrosa) [inline peepholelib/peepholes/ClassDependentClassifiers/tgmm.py]: do you think this function will ever be used at experiment level? otherwise you can just inline it into `save()`

File: peepholelib/tests/random_pr/random_pr.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import torch
+
```
Comment (lsrosa) [inline peepholelib/tests/random_pr/random_pr.py:2]: me like what I see

File: peepholelib/plots/confidence.py [hunk 14]
```diff
@@ -6,10 +6,12 @@
 from matplotlib.lines import Line2D
```
Comment (lsrosa) [inline peepholelib/plots/confidence.py:6]: Can you move the plottings for another PR, updating all the plottings according to the MACS paper?

**Review by lsrosa — CHANGES_REQUESTED** [review]

**Comment by lsrosa** [comment]
I would be interested in seen some numbers before merging this, specially the CDC, it seems not a largely scalable thing.

File: peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py (hunk 5 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py:27]: No i changed it to make it work for SwinB since we weren't using it. I thought to recycle it for another purpose

File: peepholelib/peepholes/ClassDependentClassifiers/tgmm.py (hunk 11 above)
Comment (lorenzoCapelli) [reply peepholelib/peepholes/ClassDependentClassifiers/tgmm.py]: Yes

File: peepholelib/plots/confidence.py (hunk 14 above)
Comment (lorenzoCapelli) [reply peepholelib/plots/confidence.py:6]: ok

File: peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py (hunk 7 above)
Comment (lorenzoCapelli) [reply peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py]: the anomaly detector part is not ready yet then i move it to another pr

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 4 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/randomProjection.py:124]: the QR decomposition is used to orthogonalize the random matrix we generated with the random process

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 3 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/randomProjection.py:68]: you need the model during for defining the size of the matrix

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 2 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/randomProjection.py]: ok

File: peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py (hunk 5 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py:27]: The architecture of SwinB is different from the one of the ViT and since it is not our method i looked for the easiest solution to apply a dim reduction that resembled the one of the original paper

File: peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py (hunk 8 above)
Comment (lorenzoCapelli) [reply peepholelib/peepholes/ClassDependentClassifiers/anomaly_detector.py]: why should if this check doesn't pass the code wouldn't reach the fit no?

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 3 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/randomProjection.py:68]: because other wise how do you get the activation size if you don't do the inference of the model?

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 4 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/randomProjection.py:124]: I decided to use an approach similar to the one used in SVD. INstead of deciding at priory the size of the corevectors and therefore of the random matrix i take a sufficiently large random matrix and then i trim it at any desired size

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 3 above)
Comment (lsrosa) [reply peepholelib/coreVectors/dimReduction/randomProjection.py:68]: Sorry, my bad, you are right.

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 4 above)
Comment (lsrosa) [reply peepholelib/coreVectors/dimReduction/randomProjection.py:124]: Yes, the idea is clear, but if the QR does not have the same property of the SVDs, doing this trimming makes no sense. So my question is: Does it have the same property of the SVDs?

File: peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py (hunk 5 above)
Comment (lsrosa) [reply peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py:27]: Maybe update the file name from "vit_token_wise_mean.py" to "swinb_token_wise_mean.py"?
I think we should find a better structure for saving the model-specific dim reductions

Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/swin_token_wise_mean.py:27]: ok

File: peepholelib/coreVectors/dimReduction/randomProjection.py (hunk 4 above)
Comment (lorenzoCapelli) [reply peepholelib/coreVectors/dimReduction/randomProjection.py:124]: the idea is that you decide a dimension that is higher than the one you need in this way you trim the corevector to the desired dimension. You could also think of use a randomic pick of the rows instead of choosing them in the order in which is saved
